### DISCUSSION TITLE
How to represent nested union?

### DISCUSSION THREAD

**Discussion by RafalFilipek** [microsoft/typespec#2349] [Q&A] [answered]
Hi,

Lets say I have 4 models:

```
model Foo {type: "FOO"}
model BarA {type: "BAR", role: "A"}
model BarB {type: "BAR", role: "B"}
model Biz {type: "BIZ"}
```

I have a problem with implementing union where response is in form:

```
@discriminator("type")
union Items {
  FOO: Foo,
  BAR: BarA | BarB, // I know this is invalid syntax
  BIZ: Biz,
}
```

My first though was to create an additional union for `BarA` and `BarB` but this gives me an error :

```
@discriminator("role")
union Bar {
  A: BarA,
  B: BarB,
}
```

```
Union variant "BAR" must be a model type.
```

I also tried with additional model that `BarA` and `BarB` and extend:

```
@discriminator("role")
model Bar { ... }
```

but also no luck.

Is there any way to represent this with `TypeSpec`?

Thanks,

Rafał

**Comment by mikekistler** [comment, marked as answer]
Here's a suggestion:
```
model Foo {type: "FOO"}
@discriminator("role")
model Bar {
  type: "BAR"
}
model BarA extends Bar {type: "BAR", role: "A"}
model BarB extends Bar {type: "BAR", role: "B"}
model Biz {type: "BIZ"}

@discriminator("type")
union Items {
  FOO: Foo,
  BAR: Bar,
  BIZ: Biz,
}
```

[Playground link](https://cadlplayground.z22.web.core.windows.net/?c=bW9kZWwgRm9vIHt0eXBlOiAiRk9PIn0KQGRpc2NyaW1pbmF0b3IoInJvbGUiKQrGL0JhciB7CiAgxzJCQVIiCn3KHEEgZXh0ZW5kc8YpyyYsIMRIOiAiQSLLMELfMMQNyjBpesojSVoifQrxAKzEHiIpCnVuaW9uIEl0ZW1z5QCuRk9POuQA6iwKICBCQVI6xG7FDElaOsRXLAp9Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

Is this what you wanted?
